libsel4/tools/cgenerator.py

```diff
@@ -0,0 +1,598 @@
+
+            Type("seL4_Uint8", 8, wordsize),
+            Type("seL4_Uint16", 16, wordsize),
+            Type("seL4_Uint32", 32, wordsize),
+            Type("seL4_Uint64", 64, wordsize, double_word=(wordsize == 32)),
+            Type("seL4_Time", 64, wordsize, double_word=(wordsize == 32)),
+            Type("seL4_Word", wordsize, wordsize),
+            Type("seL4_Bool", 1, wordsize, native_size_bits=8),
+
+            # seL4 Structures
+            BitFieldType("seL4_CapRights_t", wordsize, wordsize),
+
+            # Object types
+            CapType("seL4_CPtr", wordsize),
+            CapType("seL4_CNode", wordsize),
+            CapType("seL4_IRQHandler", wordsize),
+            CapType("seL4_IRQControl", wordsize),
+            CapType("seL4_TCB", wordsize),
+            CapType("seL4_Untyped", wordsize),
+            CapType("seL4_DomainSet", wordsize),
+            CapType("seL4_SchedContext", wordsize),
+            CapType("seL4_SchedControl", wordsize),
+        ]
+
+    def init_arch_types(self):
+        wordsize = self.wordsize
+        CapType = CGenerator.CapType
+        StructType = CGenerator.StructType
+        Type = CGenerator.Type
+        arm_smmu = [
+            CapType("seL4_ARM_SIDControl", wordsize),
+            CapType("seL4_ARM_SID", wordsize),
+            CapType("seL4_ARM_CBControl", wordsize),
+            CapType("seL4_ARM_CB", wordsize),
+        ]
+        arch_types = {
+            "aarch32": [
+                Type("seL4_ARM_VMAttributes", wordsize, wordsize),
+                CapType("seL4_ARM_Page", wordsize),
+                CapType("seL4_ARM_PageTable", wordsize),
+                CapType("seL4_ARM_PageDirectory", wordsize),
+                CapType("seL4_ARM_ASIDControl", wordsize),
+                CapType("seL4_ARM_ASIDPool", wordsize),
+                CapType("seL4_ARM_VCPU", wordsize),
+                CapType("seL4_ARM_IOSpace", wordsize),
+                CapType("seL4_ARM_IOPageTable", wordsize),
+                StructType("seL4_UserContext", wordsize * 19, wordsize),
+            ] + arm_smmu,
+
+            "aarch64": [
+                Type("seL4_ARM_VMAttributes", wordsize, wordsize),
+                CapType("seL4_ARM_Page", wordsize),
+                CapType("seL4_ARM_PageTable", wordsize),
+                CapType("seL4_ARM_PageDirectory", wordsize),
+                CapType("seL4_ARM_PageUpperDirectory", wordsize),
+                CapType("seL4_ARM_PageGlobalDirectory", wordsize),
+                CapType("seL4_ARM_VSpace", wordsize),
+                CapType("seL4_ARM_ASIDControl", wordsize),
+                CapType("seL4_ARM_ASIDPool", wordsize),
+                CapType("seL4_ARM_VCPU", wordsize),
+                CapType("seL4_ARM_IOSpace", wordsize),
+                CapType("seL4_ARM_IOPageTable", wordsize),
+                StructType("seL4_UserContext", wordsize * 36, wordsize),
+            ] + arm_smmu,
+
+            "arm_hyp": [
```

Comment:
Since `arm_hyp` is basically `aarch32`, is there a chance to map this to `aarch32` to reduce the duplication and avoid setting a bad example?

Reply:
Yes, right now we have these lines instead:
```      
# legacy alias
arch_types["arm_hyp"] = arch_types["aarch32"]
```